### PR TITLE
[mlir][vector] Constrain patterns: vector.contract -> vector.outerproduct

### PR DESCRIPTION
**Pull request by banach-space**
This patch constrains the patterns for converting `vector.contract` to
`vector.outerproduct` so that

  * the reduction dimension is _not unrolled_ if the corresponding
    dimension is scalable.

This is necessary as the current lowering is incorrect for scalable
dims. Indeed, instead of the following unrolling that is currently being
generated for `vector.contract` (K is the size  f the reduction
dimension):

```
  // K times
  %lhs = vector.extract %LHS[0]
  %rhs = vector.extract %RHS[0]
  vector.outerproduct %lhs, %rhs

  %lhs = vector.extract %LHS[1]
  %rhs = vector.extract %RHS[1]
  vector.outerproduct %lhs, %rhs

  ...
```

we should be generating a `for` loop like the following:
```
scf.for %k = 0 to K step 1
  %lhs = vector.extract LHS[%k]
  %rhs = vector.extract RHS[%k]
  vector.outerproduct %lhs, %rhs
```

However, the lowering of `vector.extract` of vector slices with dynamic
indices is incomplete and hence the implementation above wouldn't work
just yet. Instead, this patch effectively disables unrolling in case
where the generated code would be functionally incorrect (i.e. when the
reduction dimension is scalable).

In order to document unsupported cases, a dedicated test file is added:

* "vector-contract-to-outerproduct-transforms-unsupported.mlir"

This is the first patch in a series of patches that strives to update
these patterns (and to test them) for scalable vectors.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 1]
```diff
@@ -980,9 +995,15 @@ FailureOr<Value> ContractionOpLowering::lowerParallel(PatternRewriter &rewriter,
         diag << "expected lhsIndex=" << lhsIndex << " and rhsIndex=" << rhsIndex
              << " to map to the same dimension";
       });
+    // Unrolling a scalable dimension would be incorrect - bail out.
+    if (lhsType.getScalableDims()[lhsIndex])
+      return failure();
```
Comment (nicolasvasilache) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:1000]: can we notify with a message here and below ?

**Review by nicolasvasilache — APPROVED** [review]

**Review by MacDue — APPROVED** [review]
LGTM, just a small question:

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp (hunk 1 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:1000]: Yes, will do :)

File: mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir [hunk 2]
```diff
@@ -54,22 +54,29 @@ func.func @masked_extract_contract2(%arg0: vector<2x3xf32>,
   return %0 : vector<2xf32>
 }
 
-// CHECK-LABEL: func.func @masked_extract_contract4(
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-contract-to-outerproduct-transforms.mlir:57]: Copy and paste failure, sorry :( Thanks for pointing this out!

**Review by banach-space — COMMENTED** [review]
Thanks for the reviews, I will send an update shortly and then merge it.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 3]
```diff
@@ -424,9 +424,14 @@ struct UnrolledOuterProductGenerator
     return rewriter.create<arith::ExtSIOp>(loc, promotedType, v);
   }
 
-  FailureOr<Value> outerProd(Value lhs, Value rhs, Value res, int reductionSize,
+  FailureOr<Value> outerProd(Value lhs, Value rhs, Value res, VectorType lhsType,
+                             int reductionDim,
                              std::optional<Value> maybeMask = std::nullopt) {
-    assert(reductionSize > 0);
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:429]: why was this assert removed?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 4]
```diff
@@ -980,9 +995,19 @@ FailureOr<Value> ContractionOpLowering::lowerParallel(PatternRewriter &rewriter,
         diag << "expected lhsIndex=" << lhsIndex << " and rhsIndex=" << rhsIndex
              << " to map to the same dimension";
       });
+    if (lhsType.getScalableDims()[lhsIndex])
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic &diag) {
+        diag << "Unrolloing scalable dimension (lhsIndex=" << lhsIndex
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:1000]: `s/Unrolloing/Unrolling`

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 5]
```diff
@@ -980,9 +995,19 @@ FailureOr<Value> ContractionOpLowering::lowerParallel(PatternRewriter &rewriter,
...
+             << ") is not supported yet";
+      });
     dimSize = lhsType.getDimSize(lhsIndex);
   } else if (rhsIndex >= 0) {
     iterIndex = iMap[1].getDimPosition(rhsIndex);
+    if (rhsType.getScalableDims()[rhsIndex])
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic &diag) {
+        diag << "Unrolloing scalable dimension (lhsIndex=" << lhsIndex
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:1008]: this should be `rhsIndex`?

**Comment by banach-space** [comment]
@c-rhodes Thanks for the review 🙏🏻 . As this has already been merged (i.e. before you left your comments), I've uploaded a follow-up: https://github.com/llvm/llvm-project/pull/68581.

**Pull request closed without merging**